mmpy_bot/plugins/base.py

```diff
@@ -88,21 +71,185 @@ async def call_function(
+        self.plugins: Sequence[Plugin] = plugins
+
+        self.message_listeners: Dict[re.Pattern, List[MessageFunction]] = defaultdict(
+            list
+        )
+        self.webhook_listeners: Dict[re.Pattern, List[WebHookFunction]] = defaultdict(
+            list
+        )
+
+    def __iter__(self):
+        return iter(self.plugins)
+
+    def initialize_manager(self, driver: Driver, settings: Settings):
+        for plugin in self.plugins:
+            plugin.initialize(driver, self, settings)
+
+            # Register listeners for any listener functions in the plugin
+            for attribute in dir(plugin):
+                attribute = getattr(plugin, attribute)
+                if not isinstance(attribute, Function):
+                    continue
+
+                # Register this function and any potential siblings
+                for function in [attribute] + attribute.siblings:
+                    # Plugin message/webhook handlers can be decorated multiple times
+                    # resulting in multiple siblings that do not have .plugin defined
+                    # or where the relationship with the parent plugin is incorrect
+                    function.plugin = plugin
+                    if isinstance(function, MessageFunction):
+                        self.message_listeners[function.matcher].append(function)
+                    elif isinstance(function, WebHookFunction):
+                        self.webhook_listeners[function.matcher].append(function)
+                    else:
+                        raise TypeError(
+                            f"{plugin.__class__.__name__} has a function of unsupported"
+                            f" type {type(function)}."
+                        )
+
+    def _split_docstring(self, doc):
+        """Split docstring into first line (header) and full body."""
+        return (doc.split("\n", 1)[0], doc) if doc is not None else ("", "")
+
+    def _generate_plugin_help(
+        self,
+        plug_help: List[PluginHelpInfo],
+        help_type: str,
+        items: ItemsView[re.Pattern, List[Function]],
+    ):
+        """Build PluginHelpInfo objects from plugin and function information.
+
+        Returns one PluginHelpInfo instance for every listener (message or webhook)
+        """
+        for matcher, functions in items:
+            for function in functions:
+                plug_head, plug_full = self._split_docstring(function.plugin.__doc__)
+                func_head, func_full = self._split_docstring(function.docstring)
+
+                if help_type == "message":
+                    direct = function.direct_only
+                    mention = function.needs_mention
+                elif help_type == "webhook":
+                    direct = mention = False
+                else:
+                    raise NotImplementedError(f"Unknown help type: '{help_type}'")
+
+                plug_help.append(
+                    PluginHelpInfo(
```

Comment:
The class is called `PluginHelpInfo`, but it looks like this is actually a `FunctionInfo`?